---
title: "Differential expression analysis"
author: "Rylee K. Hackley"
output: 
  html_document:
    keep_md: true
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(collapse = TRUE)
```

```{r, message=F, warning=FALSE, echo=FALSE, collapse=TRUE}
# BiocManager::install("limma")
library(tidyverse)
library(ggpubr)
library(DESeq2)
library(viridis)
library(RColorBrewer)
library(pheatmap)
library(rrcov)
library(psych)
library(factoextra)
```

load ggf data info for informative gene descriptions and counts files
```{r, message=F, echo=FALSE, collapse=TRUE}
gff_df <- read_csv("../00_genome_files/GCF_000223905.1_gff.key.csv")
counts_data <- read.csv("00_deseq2_input/00_data_all.csv", row.names = 1, check.names = F)
counts_meta <- read.csv("00_deseq2_input/00_meta_all.csv", row.names = 1, check.names = F)
counts_mtx <- as.matrix(counts_data)
dim(counts_mtx)
```

### Differential Expression Analysis
```{r, warning=FALSE, message=FALSE}
dds_mydata <- DESeqDataSetFromMatrix(
  countData = counts_data,
  colData = counts_meta,
  design = ~ batch + group
)

# Estimate Size Factors to inspect the raw count data which is generated from various growth experiments, flowcells and bio reps.
dds_mydata <- estimateSizeFactors(dds_mydata)
ddsDE <- DESeq(dds_mydata, test = "Wald")

# Estimate of sample quality; a plot of per-gene dispersion estimates together with the fitted mean-dispersion relationship
plotDispEsts(ddsDE)
```

#### PCA
```{r clustering, warning=FALSE,fig.height=6, fig.width=9}
# The variance stabilizing transformation is obtained using the vst() function.
vsd <- vst(ddsDE)
assay(vsd) <- limma::removeBatchEffect(assay(vsd), vsd$batch)

# Principal Components Analysis
plotPCA(vsd, intgroup = c("genotype", "glucose"), returnData = T) -> PCA
(cbind(PCA, "pyrF"= counts_meta$pyrF, "experiment"= str_sub(counts_meta$batch, start= -4)) -> PCA)
ggplot(data = PCA) +
  geom_point(aes(x = PC1, y = PC2, fill = group, color = pyrF), size = 7, shape = 21, stroke = 2.5)+
  scale_color_manual(values = c("grey", "black")) + scale_fill_viridis(discrete = T)+
  theme_bw()+ theme(text = element_text(size = 18)) -> pca.plot
pca.plot
```

### Differential Expression Analysis (no pyrF revertants)
```{r}
counts_meta2 <- counts_meta[counts_meta$pyrF == F,]
counts_data2 <- counts_data[counts_meta2$sample]

dds.mydata2 <- DESeqDataSetFromMatrix(
  countData = counts_data2,
  colData = counts_meta2,
  design = ~ batch + group
)

## Estimate Size Factors to inspect the raw count data which is generated from various lanes and bio reps.
dds.mydata2 <- estimateSizeFactors(dds.mydata2)
ddsDE2 <- DESeq(dds.mydata2, test = "Wald")
```

### Save normalized counts
```{r, collapse=TRUE}
# Convert the raw count to the normalized count
normalized_counts <- as.data.frame(counts(ddsDE, normalized = TRUE)) %>% rownames_to_column()
write_csv(normalized_counts, file = "01_deseq2_output/normalised_counts_all.csv")

normalized_counts <- as.data.frame(counts(ddsDE2, normalized = TRUE)) %>% rownames_to_column()
write_csv(normalized_counts, file = "01_deseq2_output/normalised_counts_filtered.csv")
```

### What are the DEGs that are unique to pyrF revertant samples? What genes are DE across both analyses?
```{r, collapse=TRUE}
all <- results(ddsDE, name = "group_independent_vs_dependent", lfcThreshold = 1, alpha = 0.05)
all <- all %>% # Make a result table
  data.frame() %>%
  rownames_to_column(var = "locus_tag") %>%
  as_tibble()
all <- left_join(all, gff_df[c(3:4, 10)], by = "locus_tag")
all_sig <- all %>% filter(padj < 0.05)

no_revert <- results(ddsDE2, name = "group_independent_vs_dependent", lfcThreshold = 1, alpha = 0.05)
no_revert_all <- no_revert %>% # Make a result table
  data.frame() %>%
  rownames_to_column(var = "locus_tag") %>%
  as_tibble()
no_revert_all <- left_join(no_revert_all, gff_df[c(3:4, 10)], by = "locus_tag")
no_revert_sig <- no_revert_all %>% filter(padj < 0.05)

#export DEGs from both analyses
write_csv(all, "01_deseq2_output/group_all.csv")
write_csv(no_revert_all, "01_deseq2_output/group_filtered.csv")
```


```{r, collapse=TRUE, fig.height=6, fig.width=6}
# Genes that are unique to pryF
(all_sig[!(all_sig$locus_tag %in% no_revert_sig$locus_tag), ] -> tmp) # 1 gene: pyrF

# Genes that are unique to non-revertants
(no_revert_sig[!(no_revert_sig$locus_tag %in% all_sig$locus_tag), ] -> tmp)  # 13 genes: most in HAH_4316-21 (ABC transporter for branched chain amino acids?)

# comparing all
tmp <- inner_join(all[c(1,3,8:9)], no_revert_all[c(1,3)], by = "locus_tag", suffix = c(".pyr", ".nopyr")) 

#plot correlations
ggscatter(tmp, 
  x = "log2FoldChange.pyr", y = "log2FoldChange.nopyr",
  add = "reg.line", conf.int = TRUE,
  cor.coef = TRUE, cor.method = "pearson", size = 2.5
) +
  labs(y = "all samples", x = "no pyrF samples") +
  theme_bw() +
  theme(text = element_text(size = 18))

# of the genes that are shared, how well does expression correlate?
(tmp <- inner_join(all_sig[c(1:3,8:9)], no_revert_sig[c(1:3)], by = "locus_tag", suffix = c(".pyr", ".nopyr")))

#save DEG list
write_csv(tmp, "01_deseq2_output/group_final.csv")

ggscatter(tmp, 
  x = "log2FoldChange.pyr", y = "log2FoldChange.nopyr",
  add = "reg.line", conf.int = TRUE,
  cor.coef = TRUE, cor.method = "pearson", size = 2.5
) +
  labs(y = "all samples", x = "no pyrF samples") +
  theme_bw() +
  theme(text = element_text(size = 18)) -> correlation
correlation
```
